core/trino-main/src/main/java/io/trino/operator/TaskAllocationValidator.java

```diff
@@ -0,0 +1,81 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.trino.operator;
+
+import io.airlift.units.DataSize;
+import io.trino.memory.context.MemoryAllocationValidator;
+
+import javax.annotation.concurrent.GuardedBy;
+
+import java.util.Comparator;
+import java.util.HashMap;
+import java.util.Map;
+
+import static com.google.common.collect.ImmutableMap.toImmutableMap;
+import static io.airlift.units.DataSize.succinctBytes;
+import static io.trino.ExceededMemoryLimitException.exceededTaskMemoryLimit;
+import static java.lang.String.format;
+import static java.util.Map.Entry.comparingByValue;
+import static java.util.Objects.requireNonNull;
+
+public class TaskAllocationValidator
```

Comment:
maybe add a cooment that `TaskAllocationValidator` is required so that it can be shared between different memory allocators